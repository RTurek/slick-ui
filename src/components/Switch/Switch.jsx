import React, { Fragment } from 'react';
import { PropTypes } from 'prop-types';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import MuiSwitch from '@material-ui/core/Switch';

// https://material-ui.com/components/switches

function Switch(props) {
  const { onChange, checked, disabled, ...otherProps } = props;

  const theme = useTheme();

  const useStyles = makeStyles({
    switchBase: {
      '&$checked': {
        color: theme?.palette?.success?.main
      },
      '&$checked + $track': {
        backgroundColor: theme?.palette?.success?.main
      }
    },
    checked: {},
    track: {}
  });

  const classes = useStyles(props);

  return (
    <Fragment>
      {!disabled &&
        <MuiSwitch
          {...otherProps}
          classes={classes}
          onChange={onChange}
          checked={checked}
        />
      }
      {disabled &&
        <MuiSwitch
          {...otherProps}
          classes={classes}
          checked={checked}
          disabled
        />
      }
    </Fragment>
  );
}

Switch.propTypes = {
  classes: PropTypes.shape({}),
  /** On Change Event handler (typically makes an API call and changes the checked value) */
  onChange: PropTypes.func.isRequired,
  /** Checked sets switch to true or false / left or right */
  checked: PropTypes.bool.isRequired,
  /** Renders a disabled switch element */
  disabled: PropTypes.bool
};

Switch.defaultProps = {
  classes: {},
  disabled: false
};

export default Switch;
