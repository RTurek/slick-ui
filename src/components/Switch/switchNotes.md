#### Switch
---

This is a wrapper component for Material Switch, mostly to apply custom styling and to handle `disabled` in a more convenient way

#### Notes
---


#### Sources
---
- https://mui.com/components/switches/
