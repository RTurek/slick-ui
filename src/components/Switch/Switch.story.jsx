import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withActions } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import Switch from './Switch';

// https://material-ui.com/components/switches

const SwitchStory = (props) => {
  const [isChecked, setIsChecked] = useState(props.isCheckedByDefault);

  const handleChange = () => {
    setIsChecked(!isChecked);
  };

  return (
    <Switch
      onChange={handleChange}
      checked={isChecked}
      disabled={props.disabled}
    />
  );
};

SwitchStory.propTypes = {
  disabled: PropTypes.bool.isRequired,
  isCheckedByDefault: PropTypes.bool
};
SwitchStory.defaultProps = {
  isCheckedByDefault: true
};

export default {
  title: '1 ~ Slick / Forms / Switch',

  decorators: [
    withInfo({
      text: 'Use knobs to try out the different variants and to disable/enable.',
      header: true,
      inline: true,
      source: false
    }),
    withActions('mouseover', 'click')
  ]
};

// NOTE: There were some issues with knobs for Switch, so I just separated these out

export const enabledAndCheckedSwitch = () => <SwitchStory
  disabled={false}
  isCheckedByDefault
/>;

enabledAndCheckedSwitch.story = {
  name: 'Enabled And Checked'
};

export const enabledAndUncheckedSwitch = () => <SwitchStory
  disabled={false}
  isCheckedByDefault={false}
/>;

enabledAndUncheckedSwitch.story = {
  name: 'Enabled And Unchecked'
};

export const disabledAndCheckedSwitch = () => <SwitchStory
  disabled
  isCheckedByDefault
/>;

disabledAndCheckedSwitch.story = {
  name: 'Disabled And Checked'
};

export const disabledAndUncheckedSwitch = () => <SwitchStory
  disabled
  isCheckedByDefault={false}
/>;

disabledAndUncheckedSwitch.story = {
  name: 'Disabled And Unchecked'
};
