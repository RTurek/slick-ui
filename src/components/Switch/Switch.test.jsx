import React from 'react';
import renderer from 'react-test-renderer';
import Switch from './Switch';

test('Switch should render', () => {
  const component = (
    <Switch
      id="test"
      data-test-id="jest-test-switch-1"
      onChange={() => {}}
      checked
    />
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
