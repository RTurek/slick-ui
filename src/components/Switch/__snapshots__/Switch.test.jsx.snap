// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render 1`] = `
<span
  className="MuiSwitch-root"
>
  <span
    aria-disabled={false}
    className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-4 MuiSwitch-switchBase makeStyles-switchBase-1 MuiSwitch-colorSecondary PrivateSwitchBase-checked-5 Mui-checked makeStyles-checked-2"
    data-test-id="jest-test-switch-1"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <span
      className="MuiIconButton-label"
    >
      <input
        checked={true}
        className="PrivateSwitchBase-input-7 MuiSwitch-input"
        id="test"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="MuiSwitch-thumb"
      />
    </span>
  </span>
  <span
    className="MuiSwitch-track makeStyles-track-3"
  />
</span>
`;
